File: python/python_demo/data_structures/hashtables/custom_hash_table.py
class Node:
    """Node for linked list used in chaining for collisions."""
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None  # Pointer to the next node

class CustomHashTable:
    """Custom hash table implementation using an array of linked lists (chaining for collisions)."""

    def __init__(self, size=10):
        self.size = size
        self.table = [None] * size  # Array to store linked lists

    def _hash(self, key):
        """Generate a hash index for a given key."""
        return hash(key) % self.size

    def insert(self, key, value):
        """Insert a key-value pair into the hash table."""
        index = self._hash(key)
        if self.table[index] is None:
            self.table[index] = Node(key, value)
        else:
            # Handle collision using chaining
            current = self.table[index]
            while current:
                if current.key == key:
                    current.value = value  # Update existing key
                    return
                if current.next is None:
                    break
                current = current.next
            current.next = Node(key, value)

    def get(self, key):
        """Retrieve a value by key."""
        index = self._hash(key)
        current = self.table[index]
        while current:
            if current.key == key:
                return current.value
            current = current.next
        return None  # Key not found

    def delete(self, key):
        """Remove a key-value pair from the hash table."""
        index = self._hash(key)
        current = self.table[index]
        prev = None
        while current:
            if current.key == key:
                if prev:
                    prev.next = current.next
                else:
                    self.table[index] = current.next  # Remove head node
                return
            prev, current = current, current.next

    def display(self):
        """Print the current state of the hash table."""
        for i, node in enumerate(self.table):
            if node:
                print(f"Index {i}: ", end="")
                current = node
                while current:
                    print(f"({current.key}: {current.value})", end=" -> ")
                    current = current.next
                print("None")

def demo_custom_hash_table():
    print("=========================")
    print("Demo of custom hash table")
    print("==========================")
    custom_table = CustomHashTable()
    custom_table.insert("name", "Alice")
    custom_table.insert("age", "30")
    custom_table.insert("city", "New York")
    print("Get name:", custom_table.get("name"))  # Output: Alice
    custom_table.display()
    custom_table.delete("age")
    print("After deletion:")
    custom_table.display()

File: java/java_demo/src/data_structures/hashtables/CustomHashTable.java
package data_structures.hashtables;

class Node {
    String key;
    String value;
    Node next;

    public Node(String key, String value) {
        this.key = key;
        this.value = value;
        this.next = null;
    }
}

public class CustomHashTable {
    private int size;
    private Node[] table;

    public CustomHashTable(int size) {
        this.size = size;
        this.table = new Node[size];
    }

    private int hash(String key) {
        return Math.abs(key.hashCode()) % size;
    }

    public void insert(String key, String value) {
        int index = hash(key);
        if (table[index] == null) {
            table[index] = new Node(key, value);
        } else {
            Node current = table[index];
            while (current != null) {
                if (current.key.equals(key)) {
                    current.value = value; // Update existing key
                    return;
                }
                if (current.next == null) break;
                current = current.next;
            }
            current.next = new Node(key, value);
        }
    }

    public String get(String key) {
        int index = hash(key);
        Node current = table[index];
        while (current != null) {
            if (current.key.equals(key)) {
                return current.value;
            }
            current = current.next;
        }
        return null; // Key not found
    }

    public void delete(String key) {
        int index = hash(key);
        Node current = table[index];
        Node prev = null;

        while (current != null) {
            if (current.key.equals(key)) {
                if (prev != null) {
                    prev.next = current.next;
                } else {
                    table[index] = current.next;
                }
                return;
            }
            prev = current;
            current = current.next;
        }
    }

    public void display() {
        for (int i = 0; i < size; i++) {
            System.out.print("Index " + i + ": ");
            Node current = table[i];
            while (current != null) {
                System.out.print("(" + current.key + ": " + current.value + ") -> ");
                current = current.next;
            }
            System.out.println("None");
        }
    }

    public static void demo() {
        System.out.println("===================");
        System.out.println("Demo for Custom HashTables");
        System.out.println("===================");
        CustomHashTable customTable = new CustomHashTable(10);
        customTable.insert("name", "Alice");
        customTable.insert("age", "30");
        customTable.insert("city", "New York");

        System.out.println("Get name: " + customTable.get("name")); // Output: Alice
        customTable.display();

        customTable.delete("age");
        System.out.println("After deletion:");
        customTable.display();
    }
}

